### PR TITLE
contributing guidelines

### LINKED ISSUE #285: create CONTRIBUTING.md
describe the development environment, pull request requirements, code style, etc.

### PR DESCRIPTION
**Pull request by tomato42**
### Description

to make it easier for people to contribute to the project, add the description of the workflow and environment

### Motivation and Context

fixes #285 

### Checklist

- [x] my code follows the style requirements
- [x] the changes are also reflected in documentation and code comments
  - Readme and PR template updated
- [x] I have read the **CONTRIBUTING** document :smile: 
- [x] the added test cases cover the modified or added code
  - N/A
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed
  - N/A
- [X] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
  - N/A
- [x] new and modified scripts were ran against popular TLS implementations:
  - N/A
- [x] required version of tlslite-ng updated in requirements.txt and README.md
   - N/A
- [x] add less obvious ways to contribute:
  - recommendations for code analysers
  - wiki/documentation
- [x] note that while the OSX and windows CI is not working yet, the code should not use features available only on Linux
- [x] running linters locally, include https://github.com/cemsbr/yala
- [x] unit test guidelines/checklist
  - https://dzone.com/articles/unit-testing-checklist
  - https://www.codeproject.com/Articles/22894/Checklist-for-Unit-Unit-Integration-Testing
  - https://github.com/nosnhojn/uvm-utest/wiki/Unit-Test-Checklist
  - https://github.com/thiagonicolussi/neovim/wiki/Unit-tests
  - https://github.com/shortcut/testing
- [x] how to update local repo copy (feature branch vs `git pull --rebase`)
- [x] link to [TLS 1.3 with annotations](https://github.com/tomato42/tls13-spec/blob/notes/draft-ietf-tls-tls13.md), note that the TLS 1.3 issues likely will be there so it can be used to gather context
- [x] link the template scripts, `scripts/test-conversation.py` and `scripts/test-tls13-conversation.py`

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/300)

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -18,10 +18,10 @@ if the PR resolves an issue, please add further checkboxes that describe the
 action items or test scenarios from it
 -->
 
-- [ ] I have read the **CONTRIBUTING** document
-- [ ] my code follows the style requirements
+- [ ] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document
+- [ ] my code follows the [PEP 8](https://www.python.org/dev/peps/pep-0008/) and [PEP 257](https://www.python.org/dev/peps/pep-0257/) style guides
```
Comment (nmav) [inline .github/PULL_REQUEST_TEMPLATE.md]: That seems something to be covered by CONTRIBUTING.md rather that another tick.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -18,10 +18,10 @@ if the PR resolves an issue, please add further checkboxes that describe the
...
 - [ ] the changes are also reflected in documentation and code comments
-- [ ] the added test cases cover the modified or added code
+- [ ] the added test cases cover the modified or added code (I have checked coveralls.io report)
```
Comment (nmav) [inline .github/PULL_REQUEST_TEMPLATE.md]: Overall I find the tickbox quite large to be practical. It may introduce a significant barrier to first commit and some of the items look like documentation in CONTRIBUTION.md

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    if you want to add new feature or extend existing one. This is not
+    necessary if the change is small.
+* If you are able and want to fix the issue, fork the repository on GitHub and
+  see [Developing changes](#developing-changes) section
+
+## Ways to contribute
+
+* Look for issues in the
+  [projects](https://github.com/tomato42/tlsfuzzer/projects) page.
+  [TLS 1.3 coverage](https://github.com/tomato42/tlsfuzzer/projects/1) is the
```
Comment (nmav) [inline CONTRIBUTING.md]: Doesn't really fit into contribution guidelines. It would seem very awkward to have that two years from now.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
+# How to contribute
+
+## How to prepare
```
Comment (nmav) [inline CONTRIBUTING.md]: to prepare what? I guess a contribution. Maybe also say why is that necessary. I guess to prevent duplicate work (and in that case, you'll also need the issue to be assigned to a person)

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+1. [Fork it](https://github.com/tomato42/tlsfuzzer/fork)
+1. Clone it
+   * `git clone git@github.com:<your-github-nick>/tlsfuzzer.git`
+   * `cd tlsfuzzer`
+1. Install dependencies (as root)
+   * `pip2 install coverage ecdsa epydoc unittest2 mock`
+   * `pip3 install coverage ecdsa`
+   * `pip2 install -r build-requirements.txt`
+   * `pip3 install -r build-requirements.txt`
+1. Download tlslite-ng
```
Comment (nmav) [inline CONTRIBUTING.md]: not a comment to the guidelines per-se, but on such a work-flow a submodule fits better.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+   * `git clone https://github.com/tomato42/tlslite-ng.git .tlslite-ng`
+   * `ln -s .tlslite-ng/tlslite tlslite`
+1. Verify installation
+   * `make test`
+1. Create your feature branch (`git checkout -b my-new-feature`)
+1. Commit your changes (`git commit -am 'Add some feature'`)
+1. Run tests (`make test`)
```
Comment (nmav) [inline CONTRIBUTING.md]: Maybe write what run tests will actually do; e.g., "Run tests (make test) to verify that new functionality works against tlslite.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+  * See also [OpenStack guide](https://wiki.openstack.org/wiki/GitCommitMessages)
+    for general good ideas about git commit messages
+* Check for unnecessary whitespace with `git diff --check` before committing
+* Generally newly submitted code should have test coverage so that it can
+  be clearly shown that it works correctly.
+  * inspect results of running `coverage` reported by `make test`
+  * pull requests with code refactoring of code that does not have test
+    coverage should have test coverage of the code added first (as separate
+    commit or separate PR)
+* Assure nothing is broken by running all tests using `make test`
```
Comment (nmav) [inline CONTRIBUTING.md]: Again nothing is broken here is quite ambiguous. May worth a separate section on what is tested, and what someone who would like to extend the testsuite should do.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (tomato42) [reply .github/PULL_REQUEST_TEMPLATE.md]: yeah, that seems more like a generic requirement (code style) rather than something specific that everybody needs to be reminded of for every PR

I'll change it to references to CONTRIBUTING.md sections

File: CONTRIBUTING.md (hunk 6 above)
Comment (tomato42) [reply CONTRIBUTING.md]: it is the TL;DR section so I'd rather omit this

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (tomato42) [reply .github/PULL_REQUEST_TEMPLATE.md]: yeah, unfortunately there are multiple things to be mindful when introducing changes, and I prefer to have it in writing
